nucypher/characters/lawful.py

```diff
@@ -554,404 +541,97 @@ def get_card(self) -> 'Card':
+            self,
+            message_kits: Sequence[Union[MessageKit, PolicyMessageKit]],
+            alice_verifying_key: PublicKey, # KeyFrag signer's key
+            encrypted_treasure_map: EncryptedTreasureMap,
+            policy_encrypting_key: PublicKey, # TODO: #2792
+            ) -> List[PolicyMessageKit]:
         """
+        Attempts to retrieve reencrypted capsule fragments
+        corresponding to given message kits from Ursulas.
 
-        # The intersection of the map and our known nodes will be the known Ursulas...
-        known_treasure_ursulas = treasure_map.destinations.keys() & self.known_nodes.addresses()
+        Accepts both "clean" message kits (obtained from a side channel)
+        and "loaded" ones (with earlier retrieved capsule frags attached,
+        along with the addresses of Ursulas they were obtained from).
 
-        # while the difference will be the unknown Ursulas.
-        unknown_treasure_ursulas = treasure_map.destinations.keys() - self.known_nodes.addresses()
-
-        return unknown_treasure_ursulas, known_treasure_ursulas
-
-    def follow_treasure_map(self,
-                            treasure_map=None,
-                            block=False,
-                            new_thread=False,
-                            timeout=10,
-                            allow_missing=0):
+        Returns a list of loaded message kits corresponding to the input list,
+        with the kits containing the capsule fragments obtained during the retrieval.
+        These kits can be used as an external cache to preserve the cfrags between
+        several retrieval attempts.
         """
-        Follows a known TreasureMap.
 
-        Determines which Ursulas are known and which are unknown.
-
-        If block, will block until either unknown nodes are discovered or until timeout seconds have elapsed.
-        After timeout seconds, if more than allow_missing nodes are still unknown, raises NotEnoughUrsulas.
+        if encrypted_treasure_map.hrac in self._treasure_maps:
+            # A small optimization to avoid multiple treasure map decryptions.
+            treasure_map = self._treasure_maps[encrypted_treasure_map.hrac]
+        else:
+            # Have to decrypt the treasure map first to find out what the threshold is.
+            # Otherwise we could check the message kits for completeness right away.
+            treasure_map = self._decrypt_treasure_map(encrypted_treasure_map)
+
+        # Normalize input
+        message_kits: List[PolicyMessageKit] = [
+            message_kit.as_policy_kit(policy_encrypting_key, treasure_map.threshold)
+                if isinstance(message_kit, MessageKit) else message_kit
+            for message_kit in message_kits
+            ]
+
+        # Clear up all unrelated information from message kits before retrieval.
+        # This call may go to Porter, and we don't want to send it the ciphertexts
+        # which it cannot decrypt anyway.
```

Comment:
Just did a rebase for #2768, and noticed this comment - presumably, this comment can be removed.